plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java

```diff
@@ -0,0 +1,560 @@
+                    .collect(toImmutableList());
+
+            // we use identity column mapping; input page already contains only data columns per
+            // DataLagePageSink.getDataPage()
+            int[] identityMapping = new int[dataColumnTypes.size()];
+            for (int i = 0; i < identityMapping.length; ++i) {
+                identityMapping[i] = i;
+            }
+
+            ParquetSchemaConverter schemaConverter = new ParquetSchemaConverter(parquetTypes, dataColumnNames, false, false);
+            return new ParquetFileWriter(
+                    fileSystem.newOutputFile(path),
+                    rollbackAction,
+                    parquetTypes,
+                    dataColumnNames,
+                    schemaConverter.getMessageType(),
+                    schemaConverter.getPrimitiveTypes(),
+                    parquetWriterOptions,
+                    identityMapping,
+                    compressionCodecName,
+                    trinoVersion,
+                    false,
+                    Optional.empty(),
+                    Optional.empty());
+        }
+        catch (IOException e) {
+            throw new TrinoException(DELTA_LAKE_BAD_WRITE, "Error creating Parquet file", e);
+        }
+    }
+
+    private FileWriter createRecordFileWriter(Path path)
+    {
+        Properties schema = buildHiveSchema(dataColumnNames, dataColumnTypes);
+        return new RecordFileWriter(
+                path,
+                dataColumnNames,
+                fromHiveStorageFormat(PARQUET),
+                schema,
+                PARQUET.getEstimatedWriterMemoryUsage(),
+                conf,
+                typeManager,
+                DateTimeZone.UTC,
+                session);
+    }
+
+    private Page getDataPage(Page page)
+    {
+        Block[] blocks = new Block[dataColumnInputIndex.length];
+        for (int i = 0; i < dataColumnInputIndex.length; i++) {
+            int dataColumn = dataColumnInputIndex[i];
+            blocks[i] = page.getBlock(dataColumn);
+        }
+        return new Page(page.getPositionCount(), blocks);
+    }
+
+    private static Page extractColumns(Page page, int[] columns)
+    {
+        Block[] blocks = new Block[columns.length];
+        for (int i = 0; i < columns.length; i++) {
+            int dataColumn = columns[i];
+            blocks[i] = page.getBlock(dataColumn);
+        }
+        return new Page(page.getPositionCount(), blocks);
+    }
+
+    protected abstract void processSynthesizedColumn(DeltaLakeColumnHandle column);
+
+    protected abstract void addSpecialColumns(
+            List<DeltaLakeColumnHandle> inputColumns,
+            ImmutableList.Builder<DeltaLakeColumnHandle> dataColumnHandles,
+            ImmutableList.Builder<Integer> dataColumnsInputIndex,
+            ImmutableList.Builder<String> dataColumnNames,
+            ImmutableList.Builder<Type> dataColumnTypes);
+
+    protected abstract String getRootTableLocation();
```

Comment:
Nitpick: I prefer abstract methods at the top, right under the constructor. I don't think this is a repo style convention though.

Reply:
I can do that